### PR TITLE
Show the lastTag and the noOfAdditionalCommits

### PR DESCRIPTION
**Pull request by luxzeitlos**
To implement something like `git describe` and `git describe --tags` we need not only the tag of the current commit but the tag of the nearest commit in the chain of ancestors with a tag. Additionally we need to know how many additional commits this ancestor is away. [This is how it works](https://git-scm.com/docs/git-describe#_description).

This is the basis to solve [this issue](https://github.com/EmberSherpa/ember-cli-app-version/issues/49).

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -228,6 +250,10 @@ module.exports = function(gitPath) {
       if (tag) {
         result.tag = tag;
       }
+
+      var lastTagInfo = findLastTag(gitPathInfo.commonGitDir, result.sha);
+      result.lastTag = lastTagInfo.tag;
+      result.noOfAdditionalCommits = lastTagInfo.noOfAdditionalCommits;
```
Comment (hjdivad) [inline index.js]: maybe `commitsSinceLastTag` would be more clear?

File: index.js [hunk 2]
```diff
@@ -149,6 +149,26 @@ function findTag(gitPath, sha) {
   return tags.length ? tags[0] : false;
 }
 
+function findLastTag(gitPath, sha, noOfAdditionalCommits) {
+  noOfAdditionalCommits = noOfAdditionalCommits || 0;
+  var tag = findTag(gitPath, sha);
```
Comment (hjdivad) [inline index.js]: This is going to read and scan `.git/packed-refs` for each commit from `HEAD` to whatever has a tag, or all commits if there are no tags.

If we're going to always run `findLastTag`, we may want to cache `packed-refs`.  As I recall `git-repo-info` winds up getting called a lot in certain builds.  Is that right @rwjblue?

File: tests/fixtures/tag-on-parent/dot-git/COMMIT_EDITMSG [hunk 3]
```diff
@@ -0,0 +1,4 @@
+second commit without tag
+# Please enter the commit message for your changes. Lines starting
+# with '#' will be ignored, and an empty message aborts the commit.
+# On branch master
```
Comment (hjdivad) [inline tests/fixtures/tag-on-parent/dot-git/COMMIT_EDITMSG]: i don't think you intended to include this file

File: tests/fixtures/tag-on-parent/dot-git/config [hunk 4]
```diff
@@ -0,0 +1,7 @@
+[core]
+	repositoryformatversion = 0
+	filemode = true
+	bare = false
+	logallrefupdates = true
+	ignorecase = true
+	precomposeunicode = true
```
Comment (hjdivad) [inline tests/fixtures/tag-on-parent/dot-git/config]: i don't think you intended to include this file

File: tests/fixtures/tag-on-parent/dot-git/description [hunk 5]
```diff
@@ -0,0 +1 @@
+Unnamed repository; edit this file 'description' to name the repository.
```
Comment (hjdivad) [inline tests/fixtures/tag-on-parent/dot-git/description]: i don't think you intended to include this file

File: tests/fixtures/tag-on-parent/dot-git/hooks/applypatch-msg.sample [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+# The hook should exit with non-zero status after issuing an
+# appropriate message if it wants to stop the commit.  The hook is
+# allowed to edit the commit message file.
+#
+# To enable this hook, rename this file to "applypatch-msg".
+
+. git-sh-setup
+commitmsg="$(git rev-parse --git-path hooks/commit-msg)"
+test -x "$commitmsg" && exec "$commitmsg" ${1+"$@"}
+:
```
Comment (hjdivad) [inline tests/fixtures/tag-on-parent/dot-git/hooks/applypatch-msg.sample]: i don't think you intended to include this file

File: tests/fixtures/tag-on-parent/dot-git/hooks/commit-msg.sample [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+# SOB=$(git var GIT_AUTHOR_IDENT | sed -n 's/^\(.*>\).*$/Signed-off-by: \1/p')
+# grep -qs "^$SOB" "$1" || echo "$SOB" >> "$1"
+
+# This example catches duplicate Signed-off-by lines.
+
+test "" = "$(grep '^Signed-off-by: ' "$1" |
+	 sort | uniq -c | sed -e '/^[ 	]*1[ 	]/d')" || {
+	echo >&2 Duplicate Signed-off-by lines.
+	exit 1
+}
```
Comment (hjdivad) [inline tests/fixtures/tag-on-parent/dot-git/hooks/commit-msg.sample]: i don't think you intended to include this file

File: tests/fixtures/tag-on-parent/dot-git/hooks/post-update.sample [hunk 8]
```diff
@@ -0,0 +1,8 @@
+#!/bin/sh
+#
+# An example hook script to prepare a packed repository for use over
+# dumb transports.
+#
+# To enable this hook, rename this file to "post-update".
+
+exec git update-server-info
```
Comment (hjdivad) [inline tests/fixtures/tag-on-parent/dot-git/hooks/post-update.sample]: i don't think you intended to include this file

File: tests/fixtures/tag-on-parent/dot-git/hooks/pre-applypatch.sample [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+#
+# The hook should exit with non-zero status after issuing an
+# appropriate message if it wants to stop the commit.
+#
+# To enable this hook, rename this file to "pre-applypatch".
+
+. git-sh-setup
+precommit="$(git rev-parse --git-path hooks/pre-commit)"
+test -x "$precommit" && exec "$precommit" ${1+"$@"}
+:
```
Comment (hjdivad) [inline tests/fixtures/tag-on-parent/dot-git/hooks/pre-applypatch.sample]: i don't think you intended to include this file

File: tests/fixtures/tag-on-parent/dot-git/hooks/pre-commit.sample [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+
+If you know what you are doing you can disable this check using:
+
+  git config hooks.allownonascii true
+EOF
+	exit 1
+fi
+
+# If there are whitespace errors, print the offending file names and fail.
+exec git diff-index --check --cached $against --
```
Comment (hjdivad) [inline tests/fixtures/tag-on-parent/dot-git/hooks/pre-commit.sample]: i don't think you intended to include this file

File: tests/fixtures/tag-on-parent/dot-git/hooks/pre-push.sample [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+		commit=`git rev-list -n 1 --grep '^WIP' "$range"`
+		if [ -n "$commit" ]
+		then
+			echo >&2 "Found WIP commit in $local_ref, not pushing"
+			exit 1
+		fi
+	fi
+done
+
+exit 0
```
Comment (hjdivad) [inline tests/fixtures/tag-on-parent/dot-git/hooks/pre-push.sample]: i don't think you intended to include this file

File: tests/fixtures/tag-on-parent/dot-git/hooks/pre-rebase.sample [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+	git rev-list ^master ^topic next
+	git rev-list ^master        next
+
+	if these match, topic has not merged in next at all.
+
+To compute (2):
+
+	git rev-list master..topic
+
+	if this is empty, it is fully merged to "master".
```
Comment (hjdivad) [inline tests/fixtures/tag-on-parent/dot-git/hooks/pre-rebase.sample]: i don't think you intended to include this file

File: tests/fixtures/tag-on-parent/dot-git/hooks/pre-receive.sample [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+		case "$value" in
+		echoback=*)
+			echo "echo from the pre-receive-hook: ${value#*=}" >&2
+			;;
+		reject)
+			exit 1
+		esac
+		i=$((i + 1))
+	done
+fi
```
Comment (hjdivad) [inline tests/fixtures/tag-on-parent/dot-git/hooks/pre-receive.sample]: i don't think you intended to include this file

File: tests/fixtures/tag-on-parent/dot-git/hooks/prepare-commit-msg.sample [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+# ,|template,)
+#   /usr/bin/perl -i.bak -pe '
+#      print "\n" . `git diff --cached --name-status -r`
+#	 if /^#/ && $first++ == 0' "$1" ;;
+
+  *) ;;
+esac
+
+# SOB=$(git var GIT_AUTHOR_IDENT | sed -n 's/^\(.*>\).*$/Signed-off-by: \1/p')
+# grep -qs "^$SOB" "$1" || echo "$SOB" >> "$1"
```
Comment (hjdivad) [inline tests/fixtures/tag-on-parent/dot-git/hooks/prepare-commit-msg.sample]: i don't think you intended to include this file

File: tests/fixtures/tag-on-parent/dot-git/hooks/update.sample [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+		;;
+	*)
+		# Anything else (is there anything else?)
+		echo "*** Update hook: unknown type of update to ref $refname of type $newrev_type" >&2
+		exit 1
+		;;
+esac
+
+# --- Finished
+exit 0
```
Comment (hjdivad) [inline tests/fixtures/tag-on-parent/dot-git/hooks/update.sample]: i don't think you intended to include this file

File: tests/fixtures/tag-on-parent/dot-git/info/exclude [hunk 16]
```diff
@@ -0,0 +1,6 @@
+# git ls-files --others --exclude-from=.git/info/exclude
+# Lines that start with '#' are comments.
+# For a project mostly in C, the following would be a good set of
+# exclude patterns (uncomment them if you want to use them):
+# *.[oa]
+# *~
```
Comment (hjdivad) [inline tests/fixtures/tag-on-parent/dot-git/info/exclude]: i don't think you intended to include this file

File: tests/fixtures/tag-on-parent/dot-git/logs/HEAD [hunk 17]
```diff
@@ -0,0 +1,2 @@
+0000000000000000000000000000000000000000 e66f7ec2da3b5d06f0fe845c4fbc87247efacf62 Lukas Kohler <lukas.kohler@ontheblueplanet.com> 1507946495 +0200	commit (initial): first commit
+e66f7ec2da3b5d06f0fe845c4fbc87247efacf62 fb26504da0ed5cd9ed366f7428c06a8433fd76e6 Lukas Kohler <lukas.kohler@ontheblueplanet.com> 1507946563 +0200	commit: second commit without tag
```
Comment (hjdivad) [inline tests/fixtures/tag-on-parent/dot-git/logs/HEAD]: i don't think you intended to include this file

File: tests/fixtures/tag-on-parent/dot-git/logs/refs/heads/master [hunk 18]
```diff
@@ -0,0 +1,2 @@
+0000000000000000000000000000000000000000 e66f7ec2da3b5d06f0fe845c4fbc87247efacf62 Lukas Kohler <lukas.kohler@ontheblueplanet.com> 1507946495 +0200	commit (initial): first commit
+e66f7ec2da3b5d06f0fe845c4fbc87247efacf62 fb26504da0ed5cd9ed366f7428c06a8433fd76e6 Lukas Kohler <lukas.kohler@ontheblueplanet.com> 1507946563 +0200	commit: second commit without tag
```
Comment (hjdivad) [inline tests/fixtures/tag-on-parent/dot-git/logs/refs/heads/master]: i don't think you intended to include this file

File: tests/fixtures/tag-on-parent/dot-git/objects/e6/6f7ec2da3b5d06f0fe845c4fbc87247efacf62 [hunk 19]
```diff
@@ -0,0 +1,2 @@
+x��A
+�0E]���&i"��I:��i#���V=����?~*�:h�OR����fH�t���h����k���LI�&S��lK��Q����C��{�d��r�X�T�\�,ygԈ��9���8�]�WTo�F=
```
Comment (hjdivad) [inline tests/fixtures/tag-on-parent/dot-git/objects/e6/6f7ec2da3b5d06f0fe845c4fbc87247efacf62]: i don't think you intended to include this file

File: tests/fixtures/tag-on-parent/dot-git/objects/fb/26504da0ed5cd9ed366f7428c06a8433fd76e6 [hunk 20]
```diff
@@ -0,0 +1,2 @@
+x��A��0Eg�Sx��<�qS	!�p��q���A�ק�a��K��I������e�*P�>a�Ys��Q��>ƚU�Gr���l�̵S�%r(��F
+B5K�<uZ�T�.�6���==�҆Q8�����f4��>�4��M'���ā�C��u���o�{����^��p5����%&Y�
```
Comment (hjdivad) [inline tests/fixtures/tag-on-parent/dot-git/objects/fb/26504da0ed5cd9ed366f7428c06a8433fd76e6]: i don't think you intended to include this file

File: tests/index.js [hunk 21]
```diff
@@ -121,7 +121,34 @@ describe('git-repo-info', function() {
...
+        abbreviatedSha: 'fb26504da0',
+        tag: null,
+        committer: 'Lukas Kohler <lukas.kohler@ontheblueplanet.com>',
+        committerDate: '2017-10-14T02:02:43.000Z',
+        author: 'Lukas Kohler <lukas.kohler@ontheblueplanet.com>',
+        authorDate: '2017-10-14T02:02:43.000Z',
+        commitMessage: 'second commit without tag',
+        root: repoRoot,
+        parents: [
+          'e66f7ec2da3b5d06f0fe845c4fbc87247efacf62'
```
Comment (hjdivad) [inline tests/index.js:148]: maybe add a test to handle the merge commit case (ie multiple parents) since it looks like you already added code to make that work.  What do you think?

**Review by hjdivad — CHANGES_REQUESTED** [review]

**Comment by hjdivad** [comment]
thanks for working on this @luxferresum; LMK if any of my comments are unclear

File: tests/fixtures/tag-on-parent/dot-git/objects/fb/26504da0ed5cd9ed366f7428c06a8433fd76e6 (hunk 20 above)
Comment (luxzeitlos) [reply tests/fixtures/tag-on-parent/dot-git/objects/fb/26504da0ed5cd9ed366f7428c06a8433fd76e6]: actually this file is required!

File: tests/index.js (hunk 21 above)
Comment (luxzeitlos) [reply tests/index.js:148]: I've added a test for the simple case where just one path has a tag. However its not clear what should happen if both paths have a tag (on the same distance). However I dont think we have to figure out all theese edge cases. For now I think its OK to not guarantee a specific behavior in that case.

**Comment by luxzeitlos** [comment]
Thanks for the review @hjdivad! I hope I was able to address all your concerns.

File: tests/index.js (hunk 21 above)
Comment (hjdivad) [reply tests/index.js:148]: @luxferresum yes I agree it's really not clear what the right answer is there.

**Review by hjdivad — APPROVED** [review]
looks great, thanks @luxferresum

**Comment by hjdivad** [comment]
@rwjblue tests are failing due to arrow fn.

@luxferresum could fix, or we could bump major and drop support for ancient node.

your preference? (you already know mine...)

**Comment by rwjblue** [comment]
Lets update `.travis.yml` to run `4`, `6`, `8`, and `9` and `package.json`'s `engine` to be `>= 4` (in a separate commit)...

**Comment by hjdivad** [comment]
awesome, thanks so much for the feature @luxferresum!

Looking forward to seeing https://github.com/EmberSherpa/ember-cli-app-version/issues/49

Looks like a good feature to add 👍

**Comment by hjdivad** [comment]
released as v2.0.0
